### ISSUE TITLE
Constructing IANAZones has huge memory impact on nodejs

### ISSUE THREAD

**Issue by sechel** [moment/luxon#923]
I have not dug deep enough to really get to the bottom of the issue here but this should serve as a warning to all that use Luxon on nodejs along with timezones. I have an algorithm that uses timezones in its core loop. The first version reads like this:
```ts
const zone = new IANAZone(asset.timezone);
const assetHour = data.date.setZone(zone);
if (assetHour.hour !== 23) { return; }
```
which produced the following memory graph on the server (two instances running, scale is in MB)
![image](https://user-images.githubusercontent.com/4698832/115369477-6da51a80-a1c8-11eb-8e6a-16702489a234.png)
So each run consumed roughly 1Gb of memory that could not be freed and eventually destabilized the whole stack.
The new version that avoids using instances of `IANAZone` seems to avoid a crucial memory issue somewhere hidden in the `IANAZone` constructor or in the calls to `setZone`:
```ts
const assetHour = data.date.setZone(asset.timezone);
if (assetHour.hour !== 23) { return; }
```
This version does not consume any noteworthy memory anymore. After redeployment at `10:45` note the small bump at `10:55` that belongs to the invocation.
![image](https://user-images.githubusercontent.com/4698832/115370381-4dc22680-a1c9-11eb-80ae-feb4cffeb52b.png)
I find this stunning how two tiny seemingly equivalent pieces code can have such a huge difference in memory consumption.

**Comment by icambron** [comment]
Yeah, the Intl API is just brutal in Chrome, both in terms of memory and CPU. We cache the Intl objects we create because it's so slow, which is why the memory is ratcheting up. The best I can offer is that we could add the ability to control the cache, at least on/off.

**Comment by diesieben07** [comment]
If I understand this correctly, the difference between the two pieces of code is that the first one uses the `IANAZone` constructor directly. The 2nd piece of code still uses `IANAZone` indirectly via `setZone`, but that uses `IANAZone.create`, which implements the cache.
As such the problem doesn't even seem to be the Intl API in this case but just the sheer number of `IANAZone` instances.
The `IANAZone` constructor should be deprecated and marked private in favor of the caching factory function, just like other classes in Luxon behave.

**Comment by manqingzhou** [comment]
Just re-ran into this from the other direction (via `IANAZone.create`/`ianaZoneCache` in `src/zones/IANAZone.js`, not the `Intl` formatter caches discussed above), and wanted to leave a data point since this is still open.

`ianaZoneCache` is a plain `Map` with no capacity bound, no LRU, no TTL, keyed by the raw zone-name string passed to `IANAZone.create`. It caches an entry for *every distinct string* reaching it — including strings that aren't valid IANA zone names — since validation (`isValidZone`) happens after the cache entry already exists. `resetCache()` exists but its own docstring says it's "only ... necessary in testing scenarios," so nothing calls it in normal operation.

Zone names reach `IANAZone.create` from ordinary parsing entry points via `extractIANAZone` (`src/impl/regexParser.js`), which is wired into `DateTime.fromISO`, `fromSQL`, and the `z` token of `fromFormat` — e.g. a bracket zone suffix like `2024-01-01T00:00:00[Some/Zone]` (a common IANA-in-ISO convention). `ianaRegex` permits up to 256-character segments, so the effective key space is enormous.

Ran this against the actual `3.7.2` source:

```js
import { DateTime } from 'luxon';
function heapMB() { return (process.memoryUsage().heapUsed / 1024 / 1024).toFixed(1); }
console.log('start', heapMB());
for (let i = 0; i < 50000; i++) {
  DateTime.fromISO(`2024-01-01T00:00:00[Evil/${'A'.repeat(200)}${i}]`);
}
console.log('end', heapMB());
```

```
start 4.8
end 34.9
```

— growth survives GC (retained, not transient garbage), and scales linearly with the number of distinct strings supplied, same shape as what's already described above.

Given the constructor overhead is already known to be expensive here, it seems like the two cheapest independent mitigations would be: (1) cap `ianaZoneCache` (and `dtfCache`) at a fixed size with LRU eviction, and (2) validate against `IANAZone.isValidZone()` *before* inserting, so garbage/invalid names (which are unbounded by definition) don't each consume a permanent cache slot — orthogonal to whatever's decided about toggling the Intl caches.
